api/src/utils/apiUtils.ts

```diff
@@ -0,0 +1,84 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+
+export interface AzureExtensionApi {
+    /**
+     * The API version for this extension. It should be versioned separately from the extension and ideally remains backwards compatible.
+     */
+    apiVersion: string;
+}
+
+export interface GetApiOptions {
+    /**
+     * The ID of the extension requesting the API.
+     *
+     * @remarks This is used for telemetry purposes, to measure which extensions are using the API.
+     */
+    readonly extensionId?: string;
+}
+
+/**
+ * TODO: move to shared package
+ * @internal
+ */
+export namespace _apiUtils {
+    export interface AzureExtensionApiProvider {
+        /**
+         * Provides the API for an Azure Extension.
+         *
+         * @param apiVersionRange - The version range of the API you need. Any semver syntax is allowed. For example "1" will return any "1.x.x" version or "1.2" will return any "1.2.x" version
+         * @param options - Options for initializing the API. See {@link GetApiOptions}
+         * @throws - Error if a matching version is not found.
+         */
+        getApi<T extends AzureExtensionApi>(apiVersionRange: string, options?: GetApiOptions): T;
+    }
+
+    export class ExtensionNotFoundError extends Error {
+        constructor(extensionId: string) {
+            super(`Extension with id ${extensionId} not found.`);
+        }
+    }
+
+    /**
+     * Gets the exported API from the given extension id and version range.
+     *
+     * @param extensionId - The extension id to get the API from
+     * @param apiVersionRange - The version range of the API you need. Any semver syntax is allowed. For example "1" will return any "1.x.x" version or "1.2" will return any "1.2.x" version
+     * @param options - The options to pass when creating the API. If `options.extensionId` is left undefined, it's set to the caller extension id.
+     * @throws Error if extension with id is not installed.
+     */
+    export async function getAzureExtensionApi<T extends AzureExtensionApi>(context: vscode.ExtensionContext, extensionId: string, apiVersionRange: string, options?: GetApiOptions): Promise<T> {
```

Comment:
I think it makes more sense for this to stay in the `@microsoft/vscode-azext-utils` package.

Reply:
And then depend on utils as a runtime dependency? It'd be a circular dependency if this package depends on utils at runtime.

Reply:
Ah, yeah. That's fair. Maybe put it just in here, and reexport it from the utils package?

Reply:
Oh yeah, haven't thought of that, genius!

Reply:
Though, eventually it'd make sense to put the api utils and types into their own package as well.